src/builder.rs

```diff
@@ -435,17 +452,109 @@ impl Bundle<PartiallyAuthorized> {
     }
 }
 
+/// Generators for property testing.
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use rand::rngs::OsRng;
+
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+
+    //use pasta_curves::{pallas};
+
+    use crate::{
+        Note, Address,
+        address::testing::arb_address,
+        bundle::{Authorized, Bundle, Flags},
+        circuit::ProvingKey,
+        keys::{FullViewingKey, OutgoingViewingKey, SpendingKey},
+        note::testing::arb_note,
+        tree::{Anchor, MerklePath},
+        value::{
+            ValueSum, NoteValue,
+            testing::{arb_positive_note_value, MAX_MONEY}
+        },
+    };
+
+    use super::Builder;
+
+    /// An intermediate type used for construction of arbitrary
+    /// bundle values. This type is required because of a limitation
+    /// of the proptest prop_compose! macro which does not correctly 
+    /// handle polymorphic generator functions. Instead of generating
+    /// a bundle directly, we generate the bundle inputs, and then
+    /// are able to use the `build` function to construct the bundle
+    /// from these inputs, but using a `ValueSum` implementation that
+    /// is defined by the end user.
+    #[derive(Debug)]
+    pub struct ArbitraryBundleInputs {
+        sk: SpendingKey,
+        anchor: Anchor,
+        notes: Vec<Note>,
+        recipient_amounts: Vec<(Address, NoteValue)>,
+    }
+
+    impl ArbitraryBundleInputs {
+        /// Create a bundle from the set of arbitrary bundle inputs.
+        pub fn into_bundle<V: ValueSum>(self) -> Bundle<Authorized, V> {
+            let fvk = FullViewingKey::from(&self.sk);
+            let ovk = OutgoingViewingKey::from(&fvk);
+            let flags = Flags::from_parts(true, true);
+            let mut builder = Builder::new(flags, self.anchor);
+
+            for note in self.notes.into_iter() {
+                builder.add_spend(fvk.clone(), note, MerklePath).unwrap();
+            }
+
+            for (addr, value) in self.recipient_amounts.into_iter() {
+                builder.add_recipient(Some(ovk.clone()), addr, value, None).unwrap();
+            }
+
+            let mut rng = OsRng;
```

Comment:
Hmm, `OsRng` inside a proptest will make it non-deterministic, which seems like something that will confuse downstream users of this. This function also inlines `ProvingKey::build` which is a very heavy call (and should generally only be called once or infrequently per process), and calls `Builder::build` which will generate proofs (also very heavy for a proptest). What is the purpose of generating arbitrary consensus-valid `Bundle`s, rather than arbitrary structurally-valid `Bundle`s?